core/example/longrange/spme.cpp

```diff
@@ -206,88 +182,330 @@ double TPME::compute( ParticleList &particles, ParticleList &mesh, double lx,
+        auto elapsed_time_rij = end_time_rij - start_time_rij;
+        auto start_time_rii = std::chrono::high_resolution_clock::now();
+
+        // energy contribution from local cells
+        double Ur_ii = 0.0;
+        Kokkos::parallel_reduce(
+            n_cells,
+            KOKKOS_LAMBDA( const int ic, double &Ur_ic ) {
+                int ix, iy, iz;
+                cell_list.ijkBinIndex( ic, ix, iy, iz );
+
+                double contrib = 0.0;
+                for ( int ii = 0; ii < cell_list.binSize( ix, iy, iz ); ++ii )
+                {
+                    int i = cell_list.binOffset( ix, iy, iz ) + ii;
+                    double rx = r( i, 0 );
+                    double ry = r( i, 1 );
+                    double rz = r( i, 2 );
+                    double Ur_i = 0.0;
+                    double fx_i = 0.0;
+                    double fy_i = 0.0;
+                    double fz_i = 0.0;
+
+                    for ( int ij = ii + 1; ij < cell_list.binSize( ix, iy, iz );
+                          ++ij )
+                    {
+                        int j = cell_list.binOffset( ix, iy, iz ) + ij;
+                        double dx = rx - r( j, 0 );
+                        double dy = ry - r( j, 1 );
+                        double dz = rz - r( j, 2 );
+                        dx -= round( dx / lx ) * lx;
+                        dy -= round( dy / lx ) * lx;
+                        dz -= round( dz / lx ) * lx;
+                        double d = sqrt( dx * dx + dy * dy + dz * dz );
+                        double qij = q( i ) * q( j );
+                        double pij =
+                            ( d <= r_max ) * 0.5 * qij * erfc( alpha * d ) / d;
+                        double f_fact =
+                            ( d <= r_max ) * qij *
+                            ( 2.0 * sqrt( alpha / PI ) * exp( -alpha * d * d ) +
+                              erfc( sqrt( alpha ) * d ) ) /
+                            ( d * d );
+
+                        double fx = f_fact * dx;
+                        double fy = f_fact * dy;
+                        double fz = f_fact * dz;
+
+                        fx_i += fx;
+                        fy_i += fy;
+                        fz_i += fz;
+
+                        Kokkos::atomic_add( &f( j, 0 ), -fx );
+                        Kokkos::atomic_add( &f( j, 1 ), -fy );
+                        Kokkos::atomic_add( &f( j, 2 ), -fz );
+
+                        Ur_i += pij;
+                        Kokkos::atomic_add( &p( j ), pij );
                     }
+                    Kokkos::atomic_add( &p( i ), Ur_i );
+                    Kokkos::atomic_add( &f( i, 0 ), fx_i );
+                    Kokkos::atomic_add( &f( i, 1 ), fy_i );
+                    Kokkos::atomic_add( &f( i, 2 ), fz_i );
+                    contrib += 2.0 * Ur_i;
                 }
-            }
-            Ur_part *= 0.5;
-            p( idx ) += Ur_part;
-        },
-        Ur );
-    Kokkos::fence();
```

Comment:
I had been putting fences after parallel loops where I thought GPU execution may need to completely finish before moving on. I wonder if I was being too cautious here, or would a fence be a good idea?

Reply:
I have added a fence after the computations of Ur_ii, which as far as I can see could overlap with the computation of Ur_ij.